patch_rebaser/patch_rebaser.py

```diff
@@ -1,3 +1,164 @@
+#!/usr/bin/env python3
 # -*- coding: utf-8 -*-
-
 """Main module."""
+
+import configparser
+import logging
+import os
+
+from distroinfo import info, query
+from git_wrapper.repo import GitRepo
+
+LOGGER = logging.getLogger("patch_rebaser")
+
+
+def find_patches_branch(repo, remote, distgit_branch):
+    """Guess patches branch name"""
+    # Adapted from rdopkg guess.find_patches_branch
+    parts = distgit_branch.split('-')
+    while parts:
+        if 'trunk' in distgit_branch:
+            branch = '%s-trunk-patches' % '-'.join(parts)
+        else:
+            branch = '%s-patches' % '-'.join(parts)
+        LOGGER.debug("Checking if branch %s exists...", branch)
+        if repo.branch.exists(branch, remote):
+            return branch
+        parts.pop()
+    return None
+
+
+def get_downstream_distgit_branch(user):
+    """Get downstream distgit branch info from DLRN projects.ini"""
+    # projects.ini location is standard
+    DLRN_PROJECTS_INI = "/usr/local/share/dlrn/{0}/projects.ini".format(user)
+
+    config = configparser.ConfigParser()
+    config.read(os.path.expandvars(DLRN_PROJECTS_INI))
+    return config.get('downstream_driver', 'downstream_distro_branch')
+
+
+def get_patches_branch(repo, remote, user):
+    """Get the patches branch name"""
+    # Get downstream distgit branch from DLRN config
+    distgit_branch = get_downstream_distgit_branch(user)
+
+    # Guess at patches branch based on the distgit branch name
+    return find_patches_branch(repo, remote, distgit_branch)
+
+
+def parse_distro_info_path(path):
+    """Break distro_info path into repo + file"""
+    path = path.strip().rsplit("/", 1)
+    info_repo = path[0]
+    info_file = path[1]
+    remote = False
+
+    if info_repo.startswith("http"):
+        remote = True
+
+    return info_file, info_repo, remote
+
+
+def get_distro_info(distroinfo_repo):
+    """Set up distro_info based on path"""
+    info_file, info_repo, remote = parse_distro_info_path(distroinfo_repo)
+
+    if remote:
+        di = info.DistroInfo(info_file, remote_git_info=info_repo)
+    else:
+        di = info.DistroInfo(info_file, local_info=info_repo)
+
+    return di.get_info()
+
+
+def get_patches_repo(distroinfo_repo, pkg_name, key):
+    """Get URL of repo with the patches branch"""
+    distro_info = get_distro_info(distroinfo_repo)
+    pkg = query.get_package(distro_info, pkg_name)
+    return pkg[key]
+
+
+def get_rebaser_config():
+    """Return a configparser object for patch_rebaser config"""
+    # Get the config file location based on path of currently running script
+    config_file = os.path.realpath(
+        "{0}/{1}".format(
+            os.path.dirname(os.path.realpath(__file__)),
+            "patch_rebaser.ini"
+        )
+    )
+    if not os.path.exists(config_file):
+        raise Exception(
+            "Configuration file {0} not found.".format(config_file)
```

Comment:
Do we want to log an error here as well?

Reply:
Sure. Done. I wasn't sure if we wanted to keep going when there is no patches repo defined or if it's expected to be that way for some packages. I'm worried we may silently fail and do nothing, so I added a warning in that case. We'll probably want to set things up so that the script logs to a file somewhere.

Reply:
The exception gets logged anyway at the end (~L169) so I think we're okay.

Reply:
Works for me